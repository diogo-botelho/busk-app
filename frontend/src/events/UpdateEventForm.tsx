import { ChangeEvent, FormEvent, useState, useContext } from "react";
import {
  Button,
  Container,
  Form,
  Row,
  Col,
  FloatingLabel,
} from "react-bootstrap";

import { UpdateEventFormData } from "../interfaces/UpdateEventFormData";
import { Event } from "../interfaces/Event";
import { UserContext } from "../users/UserContext";

interface UpdateEventFormParams {
  event: Event;
  updateEvent: (
    event: Event,
    userId: number,
    formData: UpdateEventFormData
  ) => void;
}

/**Update Event form.
 *
 * Shows form and manages update to state on changes.
 *
 * Props:
 *  - event: event object {eventId, buskerId, title, type, coordinates{lat,lng}}
 *  - updateEvent(): function that handles updating an event.
 *
 * State:
 *  - formData: tracks data inserted into form.
 *
 * On submission:
 *  - calls updateEvent function prop
 *
 * EventCard -> UpdateEventForm
 * */

export function UpdateEventForm({ event, updateEvent }: UpdateEventFormParams) {
  const currentUser = useContext(UserContext);
  const [formData, setFormData] = useState<UpdateEventFormData>({
    title: event.title,
    type: event.type,
  });

  /** Handle form submit.
   *
   * Calls updateEvent func prop.
   */
  async function handleSubmit(evt: FormEvent<HTMLFormElement>) {
    evt.preventDefault();
    if (!currentUser) {
      console.log("Please log in to update the event.");
    } else {
      updateEvent(event, currentUser.id, formData);
    }
  }

  /** Update form data field */
  function handleChange(
    evt: ChangeEvent<HTMLInputElement & HTMLSelectElement>
  ) {
    const { name, value } = evt.target;
    setFormData((prevData) => ({ ...prevData, [name]: value }));
  }

  return (
    <Container className="UpdateEventForm">
      <h5>Select a location on the map</h5>
      <Form onSubmit={handleSubmit}>
        <Row className="justify-content-center">
          <Col xs={8} className="">
            <Form.Group className="">
              <FloatingLabel label="Title" className="mb-3">
                <Form.Control
                  id="title"
                  name="title"
                  value={formData.title}
                  onChange={handleChange}
                  placeholder="Title"
                />
              </FloatingLabel>
            </Form.Group>
            <Form.Group className="">
              <FloatingLabel label="Type">
                <Form.Select
                  id="type"
                  name="type"
                  value={formData.type}
                  onChange={handleChange}
                >
                  <option>concert</option>
                  <option>dance</option>
                  <option>something else</option>
                </Form.Select>
              </FloatingLabel>
            </Form.Group>
            <Button type="submit" className="btn-primary mt-2">
              Submit
            </Button>
          </Col>
        </Row>
      </Form>
    </Container>
  );
}
