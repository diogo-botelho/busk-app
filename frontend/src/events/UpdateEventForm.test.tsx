import React from "react";
import { render } from "@testing-library/react";
import { UpdateEventForm }  from "./UpdateEventForm";
import { MemoryRouter } from "react-router";

const demoEvent = {
    id: 1,
    buskerId: 1,
    title: "demo title",
    type: "demo type",
    coordinates: {lat: 1,lng: 1}
};

it("matches snapshot", function () {
  const { asFragment } = render(
      <MemoryRouter>
        <UpdateEventForm event={demoEvent} updateEvent={()=>{}}/>
      </MemoryRouter>,
  );
  expect(asFragment()).toMatchSnapshot();
});
