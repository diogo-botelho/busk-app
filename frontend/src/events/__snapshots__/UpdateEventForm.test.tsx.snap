// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="UpdateEventForm container"
  >
    <h5>
      Select a location on the map
    </h5>
    <form
      class=""
    >
      <div
        class="justify-content-center row"
      >
        <div
          class="col-8"
        >
          <div
            class=""
          >
            <div
              class="mb-3 form-floating"
            >
              <input
                class="form-control"
                id="title"
                name="title"
                placeholder="Title"
                value="demo title"
              />
              <label>
                Title
              </label>
            </div>
          </div>
          <div
            class=""
          >
            <div
              class="form-floating"
            >
              <select
                class="form-select"
                id="type"
                name="type"
              >
                <option>
                  concert
                </option>
                <option>
                  dance
                </option>
                <option>
                  something else
                </option>
              </select>
              <label>
                Type
              </label>
            </div>
          </div>
          <button
            class="btn-primary mt-2 btn btn-primary"
            type="submit"
          >
            Submit
          </button>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
